modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/IgniteSnapshotManager.java

```diff
@@ -0,0 +1,1783 @@
+
+                    if (log.isInfoEnabled())
+                        log.info("Submit partition processings tasks wiht partition allocated lengths: " + sctx0.partFileLengths);
+
+                    // Process binary meta
+                    futs.add(CompletableFuture.runAsync(() ->
+                            sctx0.snpRcv.receiveBinaryMeta(cctx.kernalContext()
+                                .cacheObjects()
+                                .metadataTypes()),
+                        sctx0.exec));
+
+                    // Process marshaller meta
+                    futs.add(CompletableFuture.runAsync(() ->
+                            sctx0.snpRcv.receiveMarshallerMeta(cctx.kernalContext()
+                                .marshallerContext()
+                                .getCachedMappings()),
+                        sctx0.exec));
+
+                    // Process partitions
+                    for (GroupPartitionId pair : sctx0.parts) {
+                        CacheConfiguration ccfg = cctx.cache().cacheGroup(pair.getGroupId()).config();
+                        String cacheDirName = cacheDirName(ccfg);
+                        Long length = sctx0.partFileLengths.get(pair);
+
+                        try {
+                            // Initialize empty partition file.
+                            if (length == 0) {
+                                FilePageStore filePageStore = (FilePageStore) storeMgr.getStore(pair.getGroupId(),
+                                    pair.getPartitionId());
+
+                                filePageStore.init();
+                            }
+                        }
+                        catch (IgniteCheckedException e) {
+                            throw new IgniteException(e);
+                        }
+
+                        CompletableFuture<Void> fut0 = CompletableFuture.runAsync(() -> {
+                                sctx0.snpRcv.receivePart(
+                                    getPartitionFileEx(storeMgr.workDir(), cacheDirName, pair.getPartitionId()),
+                                    cacheDirName,
+                                    pair,
+                                    length);
+
+                                // Stop partition writer.
+                                sctx0.partDeltaWriters.get(pair).partProcessed = true;
+                            },
+                            sctx0.exec)
+                            // Wait for the completion of both futures - checkpoint end, copy partition
+                            .runAfterBothAsync(sctx0.cpEndFut,
+                                () -> {
+                                    File delta = getPartionDeltaFile(cacheWorkDir(sctx0.nodeSnpDir, cacheDirName),
+                                        pair.getPartitionId());
+
+                                    sctx0.snpRcv.receiveDelta(delta, cacheDirName, pair);
+
+                                    boolean deleted = delta.delete();
```

Comment:
There can be a race between `partDeltaWriters.write()` and `delta.delete()`

Reply:
Fixed by adding read-write lock